flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerTest.java

```diff
@@ -940,6 +955,177 @@ public void testCancelWithSavepoint() throws Exception {
+				Option.apply("tm"),
+				true,
+				TestingTaskManager.class);
+
+			ActorGateway taskManager = new AkkaActorGateway(taskManagerRef, leaderId);
+
+			// Wait until connected
+			Object msg = new TestingTaskManagerMessages.NotifyWhenRegisteredAtJobManager(jobManager.actor());
+			Await.ready(taskManager.ask(msg, askTimeout), askTimeout);
+
+			JobCheckpointingSettings snapshottingSettings = new JobCheckpointingSettings(
+				Collections.singletonList(sourceVertex.getID()),
+				Collections.singletonList(sourceVertex.getID()),
+				Collections.singletonList(sourceVertex.getID()),
+				checkpointInterval,
+				3600000,
+				0,
+				Integer.MAX_VALUE,
+				ExternalizedCheckpointSettings.none(),
+				null,
+				true);
+
+			jobGraph.setSnapshotSettings(snapshottingSettings);
+
+			// Submit job graph
+			msg = new JobManagerMessages.SubmitJob(jobGraph, ListeningBehaviour.DETACHED);
+			Await.ready(jobManager.ask(msg, askTimeout), askTimeout);
+
+			// Wait for all tasks to be running
+			msg = new TestingJobManagerMessages.WaitForAllVerticesToBeRunning(jobGraph.getJobID());
+			Await.ready(jobManager.ask(msg, askTimeout), askTimeout);
+
+			// ----------------------------------------------------------------
+			// Super ugly... sorry! But this is one of the few bad options
+			// to test this here. Ideally, we would have a factory that we
+			// can set in tests as desired. But we don't. So here we go...
+			msg = new RequestExecutionGraph(jobGraph.getJobID());
+			Object result = Await.result(jobManager.ask(msg, askTimeout), askTimeout);
+
+			ExecutionGraph eg;
+			if (result instanceof ExecutionGraphFound) {
+				// Sorry...
+				eg = (ExecutionGraph) ((ExecutionGraphFound) result).executionGraph();
+			} else {
+				throw new RuntimeException("Could not access ExecutionGraph for job with "
+					+ "ID " + jobGraph.getJobID() + ". Response: " + result.toString());
+
+			}
+
+			Field field = eg.getClass().getDeclaredField("checkpointCoordinator");
+			field.setAccessible(true);
+			CheckpointCoordinator coord = (CheckpointCoordinator) field.get(eg);
+			CheckpointCoordinator spiedCoord = Mockito.spy(coord);
+			field.set(eg, spiedCoord);
+			// ----------------------------------------------------------------
+
+			// Cancel with savepoint
+			msg = new JobManagerMessages.CancelJobWithSavepoint(jobGraph.getJobID(),
+				savepointTarget.getAbsolutePath());
+			CancellationResponse cancelResp = (CancellationResponse) Await.result(
+				jobManager.ask(msg, askTimeout), askTimeout);
+
+			if (cancelResp instanceof CancellationFailure) {
+				if (enablePeriodicCheckpoints) {
+					// Verify checkpoint scheduler deactivated and reactivated.
+					// A call to start checkpoint scheduler calls stop scheduler
+					// again. Therefore, we verify two calls for stop. Since we
+					// spy (I know...) on the coordinator after the job has
+					// started, we don't count calls before spying.
+					verify(spiedCoord, times(1)).startCheckpointScheduler();
```

Comment:
Could we not re-attempt a cancel-with-savepoint? If the coordinator is shutdown it will fail; if it was restarted it should succeed (provided we adjust the failing source to only fail the first time). Then we wouldn't need the spying but would actually just test observable behavior.

Reply:
Can't we check that the submitted tasks sees another checkpoint barrier after a savepoint has been triggered? That way we would get around spying on the `CheckpointCoordinator`.